File: golang/pkg/dcid/client_test.go
package dcid

import (
	"os"
	"sync"
	"testing"
)

// ---------------------------------------------------------------------------
// NewClient validation
// ---------------------------------------------------------------------------

func TestNewClient_APIKeyRequired(t *testing.T) {
	_, err := NewClient(Config{})
	if err == nil {
		t.Fatal("expected error for missing apiKey")
	}
	if err.Error() != "apiKey is required in SDK configuration" {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestNewClient_DefaultsToProd(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	client, err := NewClient(Config{APIKey: "test-key"})
	if err != nil {
		t.Fatal(err)
	}
	if client.baseURL != "https://gateway.trustid.life/api" {
		t.Errorf("expected prod URL, got %s", client.baseURL)
	}
}

func TestNewClient_DevEnvironment(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	client, err := NewClient(Config{
		APIKey:      "test-key",
		Environment: EnvironmentDev,
	})
	if err != nil {
		t.Fatal(err)
	}
	if client.baseURL != "http://krakend.dev-external.trustid.life/api" {
		t.Errorf("expected dev URL, got %s", client.baseURL)
	}
}

func TestNewClient_TrimsTrailingSlash(t *testing.T) {
	os.Setenv("DCID_BASE_URL", "http://localhost:8080/api/")
	defer os.Unsetenv("DCID_BASE_URL")

	client, err := NewClient(Config{APIKey: "test-key"})
	if err != nil {
		t.Fatal(err)
	}
	if client.baseURL != "http://localhost:8080/api" {
		t.Errorf("expected trimmed URL, got %s", client.baseURL)
	}
}

func TestNewClient_EnvOverrideBaseURL(t *testing.T) {
	os.Setenv("DCID_BASE_URL", "http://custom:9090")
	defer os.Unsetenv("DCID_BASE_URL")

	client, err := NewClient(Config{APIKey: "test-key", Environment: EnvironmentDev})
	if err != nil {
		t.Fatal(err)
	}
	if client.baseURL != "http://custom:9090" {
		t.Errorf("expected custom URL, got %s", client.baseURL)
	}
}

func TestNewClient_ModulesInitialized(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	client, err := NewClient(Config{APIKey: "test-key"})
	if err != nil {
		t.Fatal(err)
	}
	if client.Auth == nil {
		t.Error("Auth module not initialized")
	}
	if client.Identity == nil {
		t.Error("Identity module not initialized")
	}
	if client.Identity.Encryption == nil {
		t.Error("Encryption sub-module not initialized")
	}
	if client.Identity.Issuer == nil {
		t.Error("Issuer sub-module not initialized")
	}
	if client.Identity.IPFS == nil {
		t.Error("IPFS sub-module not initialized")
	}
	if client.Identity.Verification == nil {
		t.Error("Verification sub-module not initialized")
	}
	if client.Analytics == nil {
		t.Error("Analytics module not initialized")
	}
}

func TestNewClient_CustomTimeout(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	_, err := NewClient(Config{
		APIKey:  "test-key",
		Timeout: 60_000_000_000, // 60s
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestNewClient_CustomLogger(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	_, err := NewClient(Config{
		APIKey: "test-key",
		Logger: NewConsoleLogger(false),
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestNewClient_EnableRequestLogging(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	enable := true
	_, err := NewClient(Config{
		APIKey:               "test-key",
		EnableRequestLogging: &enable,
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestNewClient_DisableRequestLogging(t *testing.T) {
	os.Unsetenv("DCID_BASE_URL")
	disable := false
	_, err := NewClient(Config{
		APIKey:               "test-key",
		Environment:          EnvironmentDev,
		EnableRequestLogging: &disable,
	})
	if err != nil {
		t.Fatal(err)
	}
}

// ---------------------------------------------------------------------------
// Token management
// ---------------------------------------------------------------------------

func TestSetTokens_And_GetTokens(t *testing.T) {
	client := &Client{}
	client.SetTokens(TokenResponse{
		AccessToken:  "access-123",
		RefreshToken: "refresh-456",
	})
	if got := client.GetAuthToken(); got != "access-123" {
		t.Errorf("expected access-123, got %s", got)
	}
	if got := client.GetRefreshToken(); got != "refresh-456" {
		t.Errorf("expected refresh-456, got %s", got)
	}
}

func TestSetAuthToken(t *testing.T) {
	client := &Client{}
	client.SetAuthToken("token-abc")
	if got := client.GetAuthToken(); got != "token-abc" {
		t.Errorf("expected token-abc, got %s", got)
	}
}

func TestSetRefreshToken(t *testing.T) {
	client := &Client{}
	client.SetRefreshToken("refresh-xyz")
	if got := client.GetRefreshToken(); got != "refresh-xyz" {
		t.Errorf("expected refresh-xyz, got %s", got)
	}
}

func TestTokensInitiallyEmpty(t *testing.T) {
	client := &Client{}
	if got := client.GetAuthToken(); got != "" {
		t.Errorf("expected empty, got %s", got)
	}
	if got := client.GetRefreshToken(); got != "" {
		t.Errorf("expected empty, got %s", got)
	}
}

func TestSetTokens_OverwritesPrevious(t *testing.T) {
	client := &Client{}
	client.SetTokens(TokenResponse{AccessToken: "a1", RefreshToken: "r1"})
	client.SetTokens(TokenResponse{AccessToken: "a2", RefreshToken: "r2"})
	if got := client.GetAuthToken(); got != "a2" {
		t.Errorf("expected a2, got %s", got)
	}
	if got := client.GetRefreshToken(); got != "r2" {
		t.Errorf("expected r2, got %s", got)
	}
}

// ---------------------------------------------------------------------------
// Thread safety
// ---------------------------------------------------------------------------

func TestTokenAccess_ThreadSafety(t *testing.T) {
	client := &Client{}
	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(3)
		go func() {
			defer wg.Done()
			client.SetTokens(TokenResponse{AccessToken: "a", RefreshToken: "r"})
		}()
		go func() {
			defer wg.Done()
			_ = client.GetAuthToken()
		}()
		go func() {
			defer wg.Done()
			_ = client.GetRefreshToken()
		}()
	}
	wg.Wait()
}

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

func TestStringPtr(t *testing.T) {
	s := stringPtr("hello")
	if *s != "hello" {
		t.Errorf("expected hello, got %s", *s)
	}
}

File: typescript/src/client.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import type { AxiosInstance } from "axios";
import {
  MOCK_ACCESS_TOKEN,
  MOCK_REFRESH_TOKEN,
  MOCK_TOKEN_RESPONSE,
} from "./__test-utils__/mock-tokens";

// Mock createHttpClient so SDK construction doesn't make real HTTP clients
vi.mock("./utils/http", () => {
  const createMock = (): AxiosInstance =>
    ({
      get: vi.fn(),
      post: vi.fn(),
      put: vi.fn(),
      patch: vi.fn(),
      delete: vi.fn(),
      request: vi.fn(),
      defaults: { headers: { common: {} } },
      interceptors: {
        request: { use: vi.fn(), eject: vi.fn(), clear: vi.fn() },
        response: { use: vi.fn(), eject: vi.fn(), clear: vi.fn() },
      },
    } as unknown as AxiosInstance);

  return {
    createHttpClient: vi.fn(() => createMock()),
  };
});

import { DCIDServerSDK } from "./client";
import { createHttpClient } from "./utils/http";

describe("DCIDServerSDK", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("constructor", () => {
    it("should create SDK with valid apiKey", () => {
      const sdk = new DCIDServerSDK({ apiKey: "test-key" });
      expect(sdk).toBeDefined();
      expect(sdk.auth).toBeDefined();
      expect(sdk.identity).toBeDefined();
      expect(sdk.analytics).toBeDefined();
    });

    it("should throw if apiKey is missing", () => {
      expect(() => new DCIDServerSDK({ apiKey: "" })).toThrow(
        "apiKey is required in SDK configuration"
      );
    });

    it("should throw if apiKey is not provided", () => {
      expect(() => new DCIDServerSDK({} as any)).toThrow(
        "apiKey is required in SDK configuration"
      );
    });

    it("should default to prod environment", () => {
      new DCIDServerSDK({ apiKey: "test-key" });

      // First call is the unauthenticated client (auth module)
      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[0]).toBe("https://gateway.trustid.life/api");
      expect(firstCall[2]).toEqual(
        expect.objectContaining({ "X-API-Key": "test-key" })
      );
    });

    it("should use dev environment when specified", () => {
      new DCIDServerSDK({ apiKey: "test-key", environment: "dev" });

      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[0]).toBe(
        "http://krakend.dev-external.trustid.life/api"
      );
    });

    it("should use custom timeout", () => {
      new DCIDServerSDK({ apiKey: "test-key", timeout: 5000 });

      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[1]).toBe(5000);
    });

    it("should merge custom headers with X-API-Key", () => {
      new DCIDServerSDK({
        apiKey: "test-key",
        defaultHeaders: { "X-Custom": "value" },
      });

      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[2]).toEqual(
        expect.objectContaining({
          "X-API-Key": "test-key",
          "X-Custom": "value",
        })
      );
    });

    it("should create unauthenticated client for auth module (no token getters)", () => {
      new DCIDServerSDK({ apiKey: "test-key" });

      // First call — unauthenticated (auth module)
      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[3]).toBeUndefined(); // getAuthToken
      expect(firstCall[4]).toBeUndefined(); // getRefreshToken
      expect(firstCall[5]).toBeUndefined(); // refreshTokenCallback
      expect(firstCall[6]).toBeUndefined(); // onTokenRefreshed
    });

    it("should create authenticated client for identity modules (with token getters)", () => {
      new DCIDServerSDK({ apiKey: "test-key" });

      // Second call — authenticated (identity modules)
      const secondCall = (createHttpClient as any).mock.calls[1];
      expect(typeof secondCall[3]).toBe("function"); // getAuthToken
      expect(typeof secondCall[4]).toBe("function"); // getRefreshToken
      expect(typeof secondCall[5]).toBe("function"); // refreshTokenCallback
      expect(typeof secondCall[6]).toBe("function"); // onTokenRefreshed
    });

    it("should create analytics client with empty baseUrl", () => {
      new DCIDServerSDK({ apiKey: "test-key" });

      // Third call — analytics client
      const thirdCall = (createHttpClient as any).mock.calls[2];
      expect(thirdCall[0]).toBe(""); // empty baseUrl
    });

    it("should initialize all identity submodules", () => {
      const sdk = new DCIDServerSDK({ apiKey: "test-key" });

      expect(sdk.identity.encryption).toBeDefined();
      expect(sdk.identity.issuer).toBeDefined();
      expect(sdk.identity.ipfs).toBeDefined();
      expect(sdk.identity.verification).toBeDefined();
    });
  });

  describe("token management", () => {
    let sdk: DCIDServerSDK;

    beforeEach(() => {
      sdk = new DCIDServerSDK({ apiKey: "test-key" });
    });

    it("should start with no tokens", () => {
      expect(sdk.getAuthToken()).toBeUndefined();
      expect(sdk.getRefreshToken()).toBeUndefined();
    });

    it("should set and get auth token", () => {
      sdk.setAuthToken(MOCK_ACCESS_TOKEN);
      expect(sdk.getAuthToken()).toBe(MOCK_ACCESS_TOKEN);
    });

    it("should set and get refresh token", () => {
      sdk.setRefreshToken(MOCK_REFRESH_TOKEN);
      expect(sdk.getRefreshToken()).toBe(MOCK_REFRESH_TOKEN);
    });

    it("should set both tokens via setTokens", () => {
      sdk.setTokens(MOCK_TOKEN_RESPONSE);
      expect(sdk.getAuthToken()).toBe(MOCK_ACCESS_TOKEN);
      expect(sdk.getRefreshToken()).toBe(MOCK_REFRESH_TOKEN);
    });

    it("should overwrite existing tokens", () => {
      sdk.setTokens(MOCK_TOKEN_RESPONSE);
      sdk.setAuthToken("new-auth-token");
      expect(sdk.getAuthToken()).toBe("new-auth-token");
      expect(sdk.getRefreshToken()).toBe(MOCK_REFRESH_TOKEN);
    });
  });

  describe("confirmOTP auto-sets tokens", () => {
    it("should auto-set tokens after confirmOTP via onTokensReceived callback", async () => {
      const sdk = new DCIDServerSDK({ apiKey: "test-key" });

      // The first createHttpClient call returns the auth module's HTTP client.
      // We need to get the mock instance used by AuthOTP.
      // Since createHttpClient is mocked, the AuthOTP got a mock http client.
      // Let's invoke confirmOTP and verify tokens are set.
      const mockHttpClient = (createHttpClient as any).mock.results[0].value;
      (mockHttpClient.post as any).mockResolvedValue({
        data: MOCK_TOKEN_RESPONSE,
      });

      await sdk.auth.confirmOTP({
        email: "user@test.com",
        otp: "123456",
      });

      expect(sdk.getAuthToken()).toBe(MOCK_ACCESS_TOKEN);
      expect(sdk.getRefreshToken()).toBe(MOCK_REFRESH_TOKEN);
    });
  });

  describe("enableRequestLogging", () => {
    it("should enable logging by default in dev", () => {
      new DCIDServerSDK({ apiKey: "test-key", environment: "dev" });

      const firstCall = (createHttpClient as any).mock.calls[0];
      // Last arg is enableRequestLogging
      expect(firstCall[8]).toBe(true);
    });

    it("should disable logging by default in prod", () => {
      new DCIDServerSDK({ apiKey: "test-key", environment: "prod" });

      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[8]).toBe(false);
    });

    it("should respect explicit enableRequestLogging override", () => {
      new DCIDServerSDK({
        apiKey: "test-key",
        environment: "prod",
        enableRequestLogging: true,
      });

      const firstCall = (createHttpClient as any).mock.calls[0];
      expect(firstCall[8]).toBe(true);
    });
  });
});
